File: Csharp/HealthTech/Estudio.cs
namespace HealtTech
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Globalization;
    using CsvHelper;
    using CsvHelper.Configuration;
    using System.Linq;

    public class Estudio
    {
        public string EstudioID { get; set; }
        public string PacienteID { get; set; }
        public string Resultado { get; set; }
        public string FechaSolicitud { get; set; }

        // Rutas de archivos
        private string rutaPacientes = @"C:\Users\Luis Angel Seoanes O\OneDrive\Escritorio\HealtTech\Csharp\Archivos\pacientes.csv";
        private string rutaEstudios = @"C:\Users\Luis Angel Seoanes O\OneDrive\Escritorio\HealtTech\Csharp\Archivos\estudios.csv";
        private string rutaEstudiosRealizados = @"C:\Users\Luis Angel Seoanes O\OneDrive\Escritorio\HealtTech\Csharp\Archivos\EstudiosRealizados.csv";

        // Constructor de la clase Estudio
        public Estudio(string IDpaciente, string IDestudio)
        {
            EstudioID = IDestudio;
            PacienteID = IDpaciente;

            // Leer archivo CSV de estudios realizados para verificar si el estudio ya existe
            using (var reader = new StreamReader(rutaEstudiosRealizados))
            using (var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)))
            {
                var estudiosRealizados = csv.GetRecords<dynamic>().ToList();
                var estudio = estudiosRealizados
                    .FirstOrDefault(e => e.IDPaciente == PacienteID && e.IDEstudio == EstudioID);

                if (estudio != null)
                {
                    // Asignar valores si ya existe
                    Resultado = estudio.Resultado;
                    FechaSolicitud = estudio.FechaSolicitud;
                }
                else
                {
                    Resultado = null;
                    FechaSolicitud = null;
                }
            }
        }

        // Método para solicitar un nuevo estudio
        public void SolicitarEstudio(string id_paciente, string id_estudio)
        {
            // Verificar si el archivo existe
            bool archivoExiste = File.Exists(rutaEstudiosRealizados);

            // Generar los datos del nuevo estudio usando un diccionario
            var nuevoEstudio = new List<Dictionary<string, object>>
            {
                new Dictionary<string, object>
                {
                    { "IDPaciente", id_paciente },
                    { "IDEstudio", id_estudio },
                    { "Resultado", "Pendiente" },
                    { "FechaSolicitud", DateTime.Now.ToString("yyyy-MM-dd") }
                }
            };

            // Leer el archivo de estudios para obtener el nombre del estudio
            using (var reader = new StreamReader(rutaEstudios))
            using (var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)))
            {
                var estudios = csv.GetRecords<dynamic>().ToList();
                var estudioEncontrado = estudios.FirstOrDefault(e => e.ID == id_estudio);

                if (estudioEncontrado != null)
                {
                    Console.WriteLine($"Estudio de tipo '{estudioEncontrado.Nombre}' solicitado para el paciente con ID {id_paciente}.");
                }
                else
                {
                    Console.WriteLine($"No se encontró el estudio con ID {id_estudio}.");
                }
            }

            // Registrar el nuevo estudio en el archivo de estudios realizados
            using (var writer = new StreamWriter(rutaEstudiosRealizados, true))
            using (var csv = new CsvWriter(writer, new CsvConfiguration(CultureInfo.InvariantCulture)))
            {
                if (!archivoExiste)
                {
                    // Escribir el encabezado si el archivo es nuevo
                    csv.WriteHeader<Dictionary<string, object>>();
                    csv.NextRecord();
                }
                // Escribir el nuevo estudio
                csv.WriteRecords(nuevoEstudio);
            }
        }

        // Método para registrar el resultado del estudio
        public void RegistrarResultado(string resultado)
        {
            // Leer los estudios realizados
            var estudiosRealizados = new List<Dictionary<string, object>>();
            using (var reader = new StreamReader(rutaEstudiosRealizados))
            using (var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)))
            {
                estudiosRealizados = csv.GetRecords<Dictionary<string, object>>().ToList();
            }

            // Actualizar el resultado del estudio solicitado
            foreach (var estudio in estudiosRealizados)
            {
                if (estudio["IDPaciente"].ToString() == PacienteID && estudio["IDEstudio"].ToString() == EstudioID)
                {
                    estudio["Resultado"] = resultado;
                    break;
                }
            }

            // Guardar los cambios de vuelta al archivo CSV
            using (var writer = new StreamWriter(rutaEstudiosRealizados))
            using (var csv = new CsvWriter(writer, new CsvConfiguration(CultureInfo.InvariantCulture)))
            {
                csv.WriteRecords(estudiosRealizados);
            }

            Console.WriteLine("Resultado registrado con éxito.");
        }
    }
}

File: python/estudio.py
import pandas as pd
import os
import csv

class Estudios:
    def __init__(self, IDpaciente, IDestudio) -> None:
        self.EstudioID = IDestudio
        self.PacienteID = IDpaciente
        self.ruta = "Archivos/EstudiosRealizados.csv"

        df = pd.read_csv(self.ruta)
        # Buscar la fila donde el ID del paciente y el ID del estudio coinciden
        EstudioRealizado = df[(df['IDpaciente'] == self.PacienteID) & (df['IDestudio'] == self.EstudioID)]
        if not EstudioRealizado.empty:
            # Asignar valores a los atributos de la instancia
            self.resultado = EstudioRealizado.iloc[0]['Resultado']
            self.FechaSolicitud = EstudioRealizado.iloc[0]['FechaSolicitud']
        else:
            self.resultado = None
            self.FechaSolicitud = None

    def solicitarEstudio(id_paciente, id_estudio):
        """Solicita un estudio para un paciente y registra la solicitud en un archivo de estudios."""
        ruta_estudiosRealizados = "Archivos/EstudiosRealizados.csv"
        ruta_Estudios = "Archivos/estudios.csv"

        # Asegurar que el archivo existe
        archivo_existe = os.path.exists(ruta_estudiosRealizados)
        
        # Generar los datos del estudio
        nuevo_estudio = {
            "IDPaciente": id_paciente,
            "IDEstudio": id_estudio,
            "Resultado": "Pendiente",
            "FechaSolicitud": pd.Timestamp.now().strftime('%Y-%m-%d')
        }
        
        # Registrar el estudio en el archivo estudiosRealizados.csv
        with open(ruta_estudiosRealizados, mode='a', newline='', encoding='utf-8') as file:
            fieldnames = ["IDPaciente", "IDEstudio", "Resultado", "FechaSolicitud"]
            writer = csv.DictWriter(file, fieldnames=fieldnames)
            
            # Escribir el encabezado si el archivo es nuevo
            if not archivo_existe:
                writer.writeheader()
            
            # Escribir el nuevo estudio
            writer.writerow(nuevo_estudio)

            dP = pd.read_csv(ruta_Estudios)

            # Asegurarte de que el id_estudio y la columna 'ID' tengan el mismo tipo (por ejemplo, ambos enteros o ambos cadenas)
            id_estudio = str(id_estudio)  # o int, según cómo sea en el archivo CSV
            dP['ID'] = dP['ID'].astype(str)  # Esto hace que los IDs sean comparables

            # Buscar el estudio por ID
            NombreEstudio = dP.loc[dP['ID'] == id_estudio, 'Nombre']

            # Verificar si se encontró el estudio
            if not NombreEstudio.empty:
                # Obtener el valor como cadena en lugar de una serie
                NombreEstudio = NombreEstudio.iloc[0]
                print(f"Estudio de tipo '{NombreEstudio}' solicitado para el paciente con ID {id_paciente}.")
            else:
                print(f"No se encontró el estudio con ID {id_estudio}.")


    def registrarResultado(self, Resultado):
        self.resultado = Resultado
        df = pd.read_csv(self.ruta)
        # Localizar la fila del paciente con el ID y el ID del estudio, y cambiar el resultado
        df.loc[(df['IDpaciente'] == self.PacienteID) & (df['IDestudio'] == self.EstudioID), 'Resultado'] = Resultado
        
        # Guardar el DataFrame actualizado de nuevo en el archivo CSV
        df.to_csv(self.ruta, index=False)
